<a href="https://colab.research.google.com/github/art-lifec/Mathematocal_Foundations/blob/main/content/algebra/examen_vragen_algebra.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Voorbeeld Examenvragen

**IT & Artificial Intelligence**

---


In [1]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.datasets import fetch_openml

np.set_printoptions(precision=4, suppress=True)
np.random.seed(42)

# MNIST laden
mnist = fetch_openml('mnist_784', version=1, as_frame=False)
X_mnist, y_mnist = mnist.data, mnist.target.astype(int)
print(f"MNIST geladen: {X_mnist.shape}")

MNIST geladen: (70000, 784)


---

## Deel A — Korte Vragen

Beantwoord elke vraag met een kort stukje code. Druk je resultaat af met `print()`.

### A1. Vectoroperaties

Gegeven `u = np.array([2, -1, 4, 3])` en `v = np.array([1, 5, -2, 0])`.

Bereken het **dot product** van `u` en `v`, en bereken de **hoek** (in graden) tussen beide vectoren.

In [3]:
u = np.array([2, -1, 4, 3])
v = np.array([1, 5, -2, 0])

# Jouw code:
# Dot product berekenen:
dot_num = np.dot(u, v)
print(f"Dot product: {dot_num}")

# Hoek berekenen:
norm_u = np.linalg.norm(u)
norm_v = np.linalg.norm(v)

# cos(θ) = (u · v) / (||u|| × ||v||)
cos_theta = dot_num / (norm_u * norm_v)

# θ = arccos(cos(θ))
theta_radialen = np.arccos(cos_theta)
theta_graden = np.degrees(theta_radialen)
print(f"Hoek: {theta_graden}°")


Dot product: -11
Hoek: 111.51018826688748°


### A2. Norm en normalisatie

Normaliseer de vector `w = np.array([3, 0, 4, 0])` zodat deze lengte 1 heeft. Print de genormaliseerde vector en **bewijs** dat de norm inderdaad 1 is.

In [6]:
w = np.array([3, 0, 4, 0])

# Jouw code:
# Norm van w
norm_w = np.linalg.norm(w)
print(f"Norm van w: {norm_w}")

# Genormaliseerde vector
w_genormaliseerd = w / norm_w
print(f"Genormaliseerded vector: {w_genormaliseerd}")

# Norm van genormaliseerde vector
print(f"Norm van genormaliseerde vector: {np.linalg.norm(w_genormaliseerd):.10f}")

# Lengte
print("De genormaliseerde vector heeft exact lengte 1.")

Norm van w: 5.0
Genormaliseerded vector: [0.6 0.  0.8 0. ]
Norm van genormaliseerde vector: 1.0000000000
De genormaliseerde vector heeft exact lengte 1.


### A3. Matrixvermenigvuldiging — shape voorspellen

Gegeven matrices met shapes: A(3×4), B(4×2), C(2×5), D(3×3).

Bepaal **zonder code uit te voeren** welke van de volgende producten geldig zijn, en wat hun resulterende shape is. Verifieer daarna met code.

1. `A @ B`
2. `B @ A`
3. `A @ B @ C`
4. `D @ A @ B`

In [7]:
# Schrijf eerst je voorspellingen als commentaar, verifieer dan met code:
# Voorspelling:
# 1. A @ B: 3x2
# 2. A @ B @ C: 3x5
# 3. D @ A @ B: 3x2

# Code:
A = np.random.rand(3, 4) #3x4
B = np.random.rand(4, 2) #4x2
C = np.random.rand(2, 5) #2x5
D = np.random.rand(3, 3) #3x3

print(f"A @ B Shape: {(A @ B).shape}")
print(f"A @ B @ C Shape: {(A @ B @ C).shape}")
print(f"D @ A @ B Shape: {(D @ A @ B).shape}")


A @ B Shape: (3, 2)
A @ B @ C Shape: (3, 5)
D @ A @ B Shape: (3, 2)


### A4. Determinant en inverteerbaarheid

Bereken de determinant van de volgende twee matrices. Welke is inverteerbaar en waarom?

```
M1 = [[6, 3],      M2 = [[4, 2],
      [4, 2]]            [7, 5]]
```

In [8]:
M1 = np.array([[6, 3], [4, 2]])
M2 = np.array([[4, 2], [7, 5]])

# Jouw code:
# Determinant M1
det_num = np.linalg.det(M1)

# Determinant M2
det_num2 = np.linalg.det(M2)

# Print()
print(f"Determinant M1: {det_num}")
print(f"Determinant M2: {det_num2}")
print("M2 is inverteerbaar omdat de determinant geen 0.0 is zoals M1.")

Determinant M1: 0.0
Determinant M2: 5.999999999999997


### A5. Stelsel oplossen

Los het volgende stelsel op met `np.linalg.solve()` en verifieer je antwoord:

```
3x + 2y - z = 1
 x -  y + 2z = 5
2x + 3y +  z = 8
```

In [9]:
# Jouw code:
# Het stelsel in matrixnotatie
A = np.array([[3, 2, -1],
              [1, -1, 2],
              [2, 3, 1]])

b = np.array([1, 5, 8])

# Oplossen met NumPy
x_oplossing = np.linalg.solve(A, b)

print(f"Oplossing: x = {x_oplossing}")
print(f"  x = {x_oplossing[0]}")
print(f"  y = {x_oplossing[1]}")
print(f"  z = {x_oplossing[2]}")
print()

# Verificatie
print("Verificatie (A @ x moet gelijk zijn aan b):")
print(f"  A @ x = {A @ x_oplossing}")
print(f"  b     = {b}")
print(f"  Gelijk? {np.allclose(A @ x_oplossing, b)}")

Oplossing: x = [0.4 1.4 3. ]
  x = 0.4000000000000002
  y = 1.3999999999999997
  z = 3.0000000000000004

Verificatie (A @ x moet gelijk zijn aan b):
  A @ x = [1. 5. 8.]
  b     = [1 5 8]
  Gelijk? True


### A6. Inverse berekenen en verifiëren

Bereken de inverse van de matrix hieronder en toon aan dat `A @ A_inv` gelijk is aan de identiteitsmatrix.

```
A = [[2, 1],
     [5, 3]]
```

In [10]:
A = np.array([[2, 1], [5, 3]])

# Jouw code:
A_inv = np.linalg.inv(A)

print("Matrix A:")
print(A)
print()
print("Inverse A⁻¹:")
print(A_inv)
print()

# Verificatie: A × A⁻¹ = I
print("Verificatie: A @ A⁻¹ =")
print(A @ A_inv)
print()
print("Verificatie: A⁻¹ @ A =")
print(A_inv @ A)
print()
print(f"Beide zijn de identiteitsmatrix? {np.allclose(A @ A_inv, np.eye(2)) and np.allclose(A_inv @ A, np.eye(2))}")

Matrix A:
[[2 1]
 [5 3]]

Inverse A⁻¹:
[[ 3. -1.]
 [-5.  2.]]

Verificatie: A @ A⁻¹ =
[[ 1.  0.]
 [-0.  1.]]

Verificatie: A⁻¹ @ A =
[[1. 0.]
 [0. 1.]]

Beide zijn de identiteitsmatrix? True


### A7. ReLU activatie

Gegeven de vector `z = np.array([-3.2, 0.0, 1.5, -0.7, 4.1, -1.0])`, pas de ReLU activatiefunctie toe **zonder** een for-loop te gebruiken. Print het resultaat.

In [11]:
z = np.array([-3.2, 0.0, 1.5, -0.7, 4.1, -1.0])

# Jouw code:
relu = np.maximum(0, z)

print(f"ReLu resultaat: {relu}")

ReLu resultaat: [0.  0.  1.5 0.  4.1 0. ]


### A8. Neuron output

Een neuron heeft gewichten `w = [0.5, -0.3, 0.8]`, bias `b = 0.2`, en ontvangt input `x = [1.0, 2.0, 0.5]`.

Bereken de output van dit neuron **met** ReLU activatie.

In [12]:
w = np.array([0.5, -0.3, 0.8])
b = 0.2
x = np.array([1.0, 2.0, 0.5])

# Jouw code:
z = np.dot(w, x) + b
print(f"Lineaire outpur z: {z}")

output = np.maximum(0, z)
print(f"Neuron output (ReLu): {output}")

Lineaire outpur z: 0.5
Neuron output (ReLu): 0.5


### A9. Batch-verwerking

Je hebt een dataset van 250 samples. Je wilt deze verwerken in batches van 32.

Bereken met code: hoeveel volledige batches krijg je, en hoeveel samples zitten in de laatste (onvolledige) batch?

In [14]:
# Jouw code:
samples = 250
batches_size = 32

# Volledige batches
full_batches = samples // batches_size

# Overgebleven batches
overgebleven = samples % batches_size

# Print()
print(f"Volledige batches: {full_batches}")
print(f"Samples laatste batch: {overgebleven}")

Volledige batches: 7
Samples laatste batch: 26


### A10. Determinant eigenschap

Gegeven een willekeurige 3×3 matrix `A` en scalar `c = 3`, verifieer dat:

**det(c · A) = c³ · det(A)**

Leg in een commentaar uit waarom de exponent 3 is.

In [15]:
# Jouw code:
A = np.random.randn(3, 3)
c = 3

left = np.linalg.det(c * A)
right = (c**3) * np.linalg.det(A)

# Print()
print(f"det(c * A): {left}")
print(f"c³ * det(A): {right}")
print(f"Eigenschap klopt: {np.isclose(left, right)}")

# De exponent is 3 omdat mijn matrix is 3x3.
# Als je een matrix vermenigvuldigt met een scalar, dan wordt elke rij vermenigvuldigd met c.


det(c * A): 35.66364453431521
c³ * det(A): 35.66364453431521
Eigenschap klopt: True


---

## Deel B — Toepassingsvragen

Deze vragen vereisen uitgebreidere code en inzicht.

---

## B1. MNIST Gelijkenisanalyse

Je krijgt 3 MNIST-afbeeldingen. Je taak is om te bepalen welke twee het meest op elkaar lijken.

**a)** Bereken de gemiddelde afbeelding van alle '3'-en en alle '8'-en in de MNIST dataset. Visualiseer beide gemiddelde afbeeldingen naast elkaar.

**b)** Schrijf een functie `cosine_similarity(a, b)` die de cosine similarity berekent tussen twee vectoren. Bereken de cosine similarity tussen de gemiddelde '3' en de gemiddelde '8'. Bereken ook de similarity van elk met de gemiddelde '1'.

**c)** Neem één concrete MNIST afbeelding van een '3' (index 10) en gebruik je cosine similarity functie om te bepalen bij welk gemiddeld cijfer (0-9) deze het best past. Komt het juiste cijfer eruit?

In [ ]:
# Jouw code voor B1:


---

## B2. Netwerklaag Bouwen en Analyseren

**a)** Implementeer een class `Layer` met:
- `__init__(self, input_size, output_size, activation='none')` — initialiseert gewichten (schaal 0.01) en biases (nullen)
- `forward(self, X)` — berekent `X @ W + b` en past de activatie toe ('none', 'relu', of 'sigmoid')

**b)** Maak een netwerk met twee lagen: 784 → 64 (ReLU) → 10 (geen activatie). Stuur de eerste 5 MNIST-afbeeldingen door het netwerk en print de output shape na elke laag.

**c)** Voeg een **softmax** functie toe die de ruwe outputs omzet naar kansen. Toon dat de kansen voor elke afbeelding sommeren tot 1. Welk cijfer voorspelt het netwerk voor elke afbeelding (met `np.argmax`)? Vergelijk met de echte labels.

In [ ]:
# Jouw code voor B2:


---

## B3. Lineaire Regressie via de Normal Equations

Een restaurant houdt bij hoeveel klanten (x) er komen en wat de omzet (y, in €) is:

| Klanten | 15 | 23 | 31 | 42 | 50 | 58 | 67 | 74 | 85 | 93 |
|---------|----|----|----|----|----|----|----|----|----|----|  
| Omzet   | 280| 420| 550| 730| 870| 1020| 1180| 1310| 1500| 1650|

**a)** Maak de design matrix `X_design` (met kolom van enen) en los het stelsel `Xᵀ X w = Xᵀ y` op voor `w = [intercept, slope]`.

**b)** Visualiseer de data als punten en de regressielijn. Zet labels op de assen en geef de vergelijking van de lijn in de titel.

**c)** Voorspel de omzet voor 40 en 100 klanten. Zijn beide voorspellingen even betrouwbaar? Leg uit.

In [ ]:
klanten = np.array([15, 23, 31, 42, 50, 58, 67, 74, 85, 93])
omzet = np.array([280, 420, 550, 730, 870, 1020, 1180, 1310, 1500, 1650])

# Jouw code voor B3:


---

## B4. Parameteranalyse van een Neuraal Netwerk

Je ontwerpt een neuraal netwerk voor MNIST-classificatie (784 inputs, 10 outputs).

**a)** Schrijf een functie `count_parameters(architecture)` die een lijst van layer-groottes ontvangt (bijv. `[784, 128, 64, 10]`) en het totaal aantal parameters (gewichten + biases) berekent. Test met de architectuur `[784, 256, 128, 64, 10]`.

**b)** Vergelijk de volgende drie architecturen qua aantal parameters:
- Breed: `[784, 512, 10]`
- Diep: `[784, 128, 64, 32, 10]`
- Heel diep: `[784, 64, 64, 64, 64, 64, 10]`

Welke heeft het meeste parameters? Welke het minste?


In [ ]:
# Jouw code voor B4:
